frame/support/src/storage/counted_map.rs

```diff
@@ -0,0 +1,314 @@
+	core::marker::PhantomData<(Key, Value, Map, Counter)>
+);
+
+impl<Key, Value, Map, Counter>
+	StorageMapT<Key, Value>
+	for CountedStorageMap<Key, Value, Map, Counter>
+where
+	Key: FullCodec,
+	Value: FullCodec,
+	Map: StorageMapT<Key, Value>,
+	Counter: StorageValueT<u32>
+{
+	type Query = Map::Query;
+	type Hasher = Map::Hasher;
+	fn module_prefix() -> &'static [u8] {
+		Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Map::storage_prefix()
+	}
+	fn from_optional_value_to_query(v: Option<Value>) -> Self::Query {
+		Map::from_optional_value_to_query(v)
+	}
+	fn from_query_to_optional_value(v: Self::Query) -> Option<Value> {
+		Map::from_query_to_optional_value(v)
+	}
+}
+
+impl<Key, Value, Map, Counter>
+	StoragePrefixedMap<Value> for
+	CountedStorageMap<Key, Value, Map, Counter>
+where
+	Value: FullCodec,
+	Map: crate::StoragePrefixedMap<Value>,
+	Counter: StorageValueT<u32>
+{
+	fn module_prefix() -> &'static [u8] {
+		Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Map::storage_prefix()
+	}
+}
+
+impl<Key, Value, Map, Counter>
+	CountedStorageMap<Key, Value, Map, Counter>
+where
+	Key: FullCodec,
+	Value: FullCodec,
+	Map: StorageMapT<Key, Value> + crate::StoragePrefixedMap<Value>,
+	Counter: StorageValueT<u32, Query=u32>,
+{
+	// Internal helper function to track the counter as a value is mutated.
+	fn mutate_counter<
+		KeyArg: EncodeLike<Key> + Clone,
+		M: FnOnce(KeyArg, F) -> R,
+		F: FnOnce(I) -> R,
+		I,
+		R,
+	>(m: M, key: KeyArg, f: F) -> R {
+		let val_existed = Map::contains_key(key.clone());
+		let res = m(key.clone(), f);
+		let val_exists = Map::contains_key(key);
+
+		if val_existed && !val_exists {
+			// Value was deleted
+			Counter::mutate(|value| value.saturating_dec());
+		} else if !val_existed && val_exists {
+			// Value was added
+			Counter::mutate(|value| value.saturating_inc());
+		}
+
+		res
+	}
+
+	/// Get the storage key used to fetch a value corresponding to a specific key.
+	pub fn hashed_key_for<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Vec<u8> {
+		Map::hashed_key_for(key)
+	}
+
+	/// Does the value (explicitly) exist in storage?
+	pub fn contains_key<KeyArg: EncodeLike<Key>>(key: KeyArg) -> bool {
+		Map::contains_key(key)
+	}
+
+	/// Load the value associated with the given key from the map.
+	pub fn get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Map::Query {
+		Map::get(key)
+	}
+
+	/// Try to get the value for the given key from the map.
+	///
+	/// Returns `Ok` if it exists, `Err` if not.
+	pub fn try_get<KeyArg: EncodeLike<Key>>(key: KeyArg) -> Result<Value, ()> {
+		Map::try_get(key)
+	}
+
+	/// Swap the values of two keys.
+	pub fn swap<KeyArg1: EncodeLike<Key>, KeyArg2: EncodeLike<Key>>(key1: KeyArg1, key2: KeyArg2) {
+		Map::swap(key1, key2)
+	}
+
+	/// Store a value to be associated with the given key from the map.
+	pub fn insert<KeyArg: EncodeLike<Key> + Clone, ValArg: EncodeLike<Value>>(key: KeyArg, val: ValArg) {
+		if !Map::contains_key(key.clone()) {
```

Comment:
```suggestion
		if !Map::contains_key(&key) {
```

`mutate_exists` could be a bit more efficient

Reply:
clone is removed, and I don't think we can optimize further